pystac/summaries.py

```diff
@@ -0,0 +1,270 @@
+            raise Exception(
+                "Could not read fields definition file at "
+                f"{FIELDS_JSON_URL} or it is invalid.\n"
+                "Try using a local fields definition file."
+            )
+        self._set_field_definitions(Summarizer._default_field_definitions)
+
+    def _set_field_definitions(self, fields: Dict[str, Any]) -> None:
+        self.summaryfields: Dict[str, SummaryStrategy] = {}
+        for name, desc in fields["metadata"].items():
+            if isinstance(desc, dict):
+                strategy_value = desc.get("summary", True)
+                try:
+                    strategy: SummaryStrategy = SummaryStrategy(strategy_value)
+                except ValueError:
+                    strategy = SummaryStrategy.DEFAULT
+                if strategy != SummaryStrategy.DONT_SUMMARIZE:
+                    self.summaryfields[name] = strategy
+            else:
+                self.summaryfields[name] = SummaryStrategy.DEFAULT
+
+    def _update_with_item(self, summaries: "Summaries", item: "Item_Type") -> None:
+        for k, v in item.properties.items():
+            if k in self.summaryfields:
+                strategy = self.summaryfields[k]
+                if strategy == SummaryStrategy.RANGE or (
+                    strategy == SummaryStrategy.DEFAULT
+                    and isinstance(v, numbers.Number)
+                    and not isinstance(v, bool)
+                ):
+                    rangesummary: Optional[RangeSummary] = summaries.get_range(k)
+                    if rangesummary is None:
+                        summaries.add(k, RangeSummary(v, v))
+                    else:
+                        rangesummary.update_with_value(v)
+                elif strategy == SummaryStrategy.ARRAY or (
+                    strategy == SummaryStrategy.DEFAULT and isinstance(v, list)
+                ):
+                    listsummary: list = summaries.get_list(k) or []
+                    if not isinstance(v, list):
+                        v = [v]
+                    for element in v:
+                        if element not in listsummary:
+                            listsummary.append(element)
+                    summaries.add(k, listsummary)
+                else:
+                    summary: list = summaries.get_list(k) or []
+                    if v not in summary:
+                        summary.append(v)
+                    summaries.add(k, summary)
+
+    def summarize(
+        self, source: Union["Collection_Type", Iterable["Item_Type"]]
+    ) -> "Summaries":
+        """Creates summaries from items"""
+        summaries = Summaries.empty()
+        if hasattr(source, "get_all_items"):
+            for item in source.get_all_items():  # type: ignore[union-attr]
+                self._update_with_item(summaries, item)
+        else:
+            for item in source:  # type: ignore[union-attr]
+                self._update_with_item(summaries, item)
```

Comment:
This satisfies the type checker without needing to ignore the error. 

```suggestion
        if isinstance(source, Collection):
            for item in source.get_all_items():
                self._update_with_item(summaries, item)
        else:
            for item in source:
                self._update_with_item(summaries, item)
```

Reply:
done

Reply:
Thinking about this...what happens if we are dealing with a Catalog instead of a Collection? It also has the `get_all_items' method. Is that a valid type of element for this? Should we extend that `isinstance` statement to check for both classes?

Reply:
It looks like [as of 1.0.0-rc.4](https://github.com/radiantearth/stac-spec/blob/master/CHANGELOG.md#v100-rc4---2021-05-11) Catalogs no longer support summaries, so I think it makes sense to restrict this to only Collections.